### PR TITLE
Parametric modular interpretation

### PR DESCRIPTION
**Pull request by robrix**
This PR migrates `semantic-analysis` over to the parametric modular abstract interpretation framework I’ve been building. It remains superficially similar, but has a better basis for the domain effect, better ergonomics via labelled effects for the store &c., and provides an operation analogous to linking, allowing out-of-order program analysis and (typically file-level) incremental updating.

### CODE REVIEW

**Review by Elpasto28 — APPROVED** [review]

**Review by Elpasto28 — APPROVED** [review]

File: semantic-analysis/src/Analysis/Reference.hs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import System.Path as Path
+import System.Path.PartClass as Path.PartClass
+
+-- Reference
+
+data Reference = Reference
+  { refPath :: Path.AbsRelFile
+  , refSpan :: Span
+  }
+  deriving (Eq, Ord, Show)
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Reference.hs:18]: Two of the fields from `File`, bundled up together so we can use them, well, all over the place. They’re already pretty widely used and I’ll add them to closures eventually too.

Comment (robrix) [reply semantic-analysis/src/Analysis/Reference.hs:18]: Incidentally, I’m not convinced that this package is really the best home for this, `File`, and similar data. I’ll figure out a better one later.

File: semantic-analysis/src/Analysis/Typecheck.hs [hunk 2]
```diff
@@ -21,154 +22,144 @@ module Analysis.Typecheck
...
+data Monotype a
+  = Var a
   | Unit
+  | Bool
+  | Int
   | String
-  | f a :-> f a
-  | Record (Map.Map Name (f a))
-  deriving (Foldable, Functor, Generic1, Traversable)
+  | Monotype a :-> Monotype a
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Typecheck.hs:57]: We no longer represent polytypes in quite the same way, but rather as a right `Type`-module, so we can recur directly here and just wrap the whole thing in some foralls binding type variables at the end. That’s actually more in line with Hindley-Damas-Milner anyway, as it’s essentially type schemes.

File: semantic-analysis/src/Analysis/Typecheck.hs [hunk 3]
```diff
@@ -21,154 +22,144 @@ module Analysis.Typecheck
...
-  deriving (Foldable, Functor, Generic1, HFunctor, RightModule, Traversable)
 
-deriving instance (Eq   a, forall a . Eq   a => Eq   (f a), Monad f) => Eq   (Polytype f a)
-deriving instance (Ord  a, forall a . Eq   a => Eq   (f a)
-                         , forall a . Ord  a => Ord  (f a), Monad f) => Ord  (Polytype f a)
-deriving instance (Show a, forall a . Show a => Show (f a))          => Show (Polytype f a)
+data Polytype a
+  = PForAll (Polytype (Maybe a))
+  | PType (Monotype a)
+  deriving (Eq, Foldable, Functor, Generic1, Ord, Show, Traversable)
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Typecheck.hs:93]: What’d I tell you? A right `Type`-module.

cf: https://twitter.com/rob_rix/status/1146602548616413184

File: semantic-analysis/src/Analysis/Typecheck.hs [hunk 4]
```diff
@@ -21,154 +22,144 @@ module Analysis.Typecheck
...
-forAlls :: (Eq a, Syntax.Has Polytype sig m, Foldable t) => t a -> m a -> m a
+abstract :: Eq a => a -> Polytype a -> Polytype (Maybe a)
+abstract n = fmap (\ a -> a <$ guard (a == n))
+
+
+forAll :: Eq a => a -> Polytype a -> Polytype a
+forAll n body = PForAll (abstract n body)
+
+forAlls :: (Eq a, Foldable t) => t a -> Polytype a -> Polytype a
 forAlls ns body = foldr forAll body ns
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Typecheck.hs:107]: This is the _I Am Not a Number, I Am a Free Variable_ approach. It doesn’t give the fastest abstraction and substitution (relative to e.g. `bound`, but a) that doesn’t actually matter at all for our purposes, and b) the smarter next move would be HOAS anyway.

File: semantic-analysis/src/Analysis/Typecheck.hs [hunk 5]
```diff
@@ -21,154 +22,144 @@ module Analysis.Typecheck
...
+          . runState @(Set.Set (Type, Type)) mempty
           . (\ m -> do
               v <- meta
               bs <- m
               v <$ for_ bs (unify v))
-          . convergeTerm 1  (A.runHeap @Addr @Type . fix (\ eval' -> runDomain eval' . fix (cacheTerm . eval)))
-
-
-data Constraint = Type :===: Type
-  deriving (Eq, Ord, Show)
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Typecheck.hs:169]: This was, alas, too cute to live.

File: semantic-core/semantic-core.cabal [hunk 6]
```diff
@@ -44,40 +44,11 @@ library
   import: common
   hs-source-dirs: src
   exposed-modules:
-    Core.Core
-    Core.Parser
-    Core.Pretty
```
Comment (robrix) [inline semantic-core/semantic-core.cabal:49]: Pouring one out.

File: semantic-core/src/Core/Expr.hs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+-- | A lambda calculus with integers, conditionals, and exceptions.
+data Expr
+  = Var Name
+  | Abs Name Expr
+  | App Expr Expr
+  | Let Name Expr Expr
+  | Lit Int
+  | If Expr Expr Expr
+  | Die String
+  deriving (Eq, Ord, Show)
```
Comment (robrix) [inline semantic-core/src/Core/Expr.hs:26]: This is a minimalistic reference calculus that exercises the evaluation interfaces fully.

File: semantic-core/src/Core/Expr.hs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+true, false :: Expr
+true  = Abs "a" (Abs "b" (Var "a"))
+false = Abs "a" (Abs "b" (Var "b"))
+
+not' :: Expr
+not' = Abs "x" (Var "x" $$ false $$ true)
+
+and', or' :: Expr
+and' = Abs "a" (Abs "b" (Var "a" $$ true    $$ Var "b"))
+or'  = Abs "a" (Abs "b" (Var "a" $$ Var "b" $$ false))
```
Comment (robrix) [inline semantic-core/src/Core/Expr.hs:45]: As seen on a blog post I wrote one time!

**Review by robrix — COMMENTED** [review]
Ready for review.

File: semantic-analysis/src/Analysis/Concrete.hs [hunk 9]
```diff
@@ -17,183 +18,162 @@
...
-data Concrete term
-  = Closure Path.AbsRelFile Span (Named (Scope () term Addr))
+data Concrete
+  = Closure Path.AbsRelFile Span (Named (Concrete -> Concrete))
   | Unit
   | Bool Bool
+  | Int Int
   | String Text
-  | Record Env
+  | Neutral Name (Snoc (Elim Concrete))
```
[the remark this answers is not in the document]
Comment (patrickt) [reply semantic-analysis/src/Analysis/Concrete.hs:48]: Can we get a bit of documentation on this declaration? It’s different enough from its surroundings that some context would be welcome.

File: semantic-analysis/src/Analysis/FlowInsensitive.hs [hunk 10]
```diff
@@ -1,78 +1,146 @@
...
-            loop x'
+-- This applies the Kleene fixed-point theorem to finitize a monotone action. cf https://en.wikipedia.org/wiki/Kleene_fixed-point_theorem
+convergeEither
+  :: Monad m
+  => (a -> m (Either b a)) -- ^ A monadic action to perform at each iteration, starting from the result of the previous iteration or from the seed value for the first iteration. Returns of 'Left' end iteration, while 'Right' begins another iteration.
+  -> a                     -- ^ An initial seed value to iterate from.
+  -> m b                   -- ^ A computation producing the least fixed point (the first value at which the actions converge).
+convergeEither f = loop
+  where
+  loop = either pure loop <=< f
```
[the remark this answers is not in the document]
Comment (patrickt) [reply semantic-analysis/src/Analysis/FlowInsensitive.hs:122]: 🙏🏻

**Review by patrickt — APPROVED** [review]
This looks terrific!

**Review by patrickt — APPROVED** [review]

**Pull request merged**